### PR TITLE
Shadow Drive integration

### PR DESCRIPTION
**Pull request by cloakd**
Implement GenesysGo Shadow-Drive Uploader

* Batch uploading (limited to 5 until QUICK support out)
* Creates & manages storage size of the Shadow drive
* SHDW balance check vs FileSize
* If shadow drive created during the process then holds in cache for next runtime
* new `StorageType` - `shadow-drive`
* new `Config` - shadowDriveAddress - Existing shadow-drive to use

### CODE REVIEW

**Comment by cloakd** [comment]
Adjusting linting etc mb!

**Comment by 21e8** [comment]
image
bold. sure this is ok and no outstanding migration?

**Comment by cloakd** [comment]
> image
> 
> bold. sure this is ok and no outstanding migration?

Yeah backwards compatibility introduced in @solana/web3 due to 1.43 pinning. 

The package is currently using TransactionBlockhashCtor which is now depreceated under new struct `TransactionCtorFields_DEPRECATED`

File: js/packages/cli/src/helpers/upload/shadow_drive.ts [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    return `${size}${denom}`;
+  }
+
+  /**
+   * Returns the url endpoint for a given drive & file
+   * @param drive
+   * @param fileName
+   */
+  toUrl(drive: string, fileName: string): string {
+    return `https://shadow-storage.genesysgo.net/${drive}/${fileName}`;
```
Comment (21e8) [inline js/packages/cli/src/helpers/upload/shadow_drive.ts:174]: pretty sure thats `https://shdw-drive.genesysgo.net/`?

**Comment by stegaBOB** [comment]
Shadow drive currently is supported in sugar. We're deprecating this js cli soon. I can take a quick look and just do a yolo merge in if you think it's good to go for now. Not going to spend a ton of time reviewing it though.

**Comment by 21e8** [comment]
I was not aware the js cli would get deprecated

**Comment by cloakd** [comment]
> Sugar is just better in every way at this point

Would of been good to know prior to building this lol!

**Comment by cloakd** [comment]
> Shadow drive currently is supported in sugar. We're deprecating this js cli soon. I can take a quick look and just do a yolo merge in if you think it's good to go for now. Not going to spend a ton of time reviewing it though.

Would of been good to know it was being depreceated before putting in the work lol

**Comment by stegaBOB** [comment]
Yeah sorry about that. Our messaging hasn't been the best. We're working on changing that. If you want to reopen this PR in the deprecated repo I can yolo merge it in

**Comment by lorisleiva** [comment]
Hi there, sorry for the late reply. I'll be closing this PR as this repo no longer contains any code and acts as a gateway to the various products and repositories provided by Metaplex. Thank you for understanding. 🙏

**Pull request closed without merging**